File: app/model/node.go
package model

import "time"

// Node 受管节点。
// 当前阶段（观测面）只做只读：主控定时拉取节点摘要用于集中展示，不代理任何写操作。
// ConnectMode 预留 direct / tunnel 两种取值，当前只实现 direct。
type Node struct {
	BaseModel
	Name        string `json:"name" gorm:"type:varchar(128);not null"`
	Addr        string `json:"addr" gorm:"type:varchar(255);not null"`        // 面板地址，含协议与端口，如 https://1.2.3.4:5470
	Entrance    string `json:"entrance" gorm:"type:varchar(255);default:''"` // 节点安全入口，节点未开启则留空
	AccessToken string `json:"-" gorm:"type:varchar(512)"`                   // 节点只读令牌，AES 加密后存储，不出接口
	ConnectMode string `json:"connectMode" gorm:"type:varchar(32);default:'direct'"`
	SkipVerify  bool   `json:"skipVerify" gorm:"default:false"` // 节点使用自签证书时跳过 TLS 校验
	IsProd      bool   `json:"isProd" gorm:"default:false"`     // 生产节点，前端高亮标记
	Sort        int    `json:"sort" gorm:"default:0"`

	// 以下字段由摘要采集任务写入，代表最近一次采集结果
	Status     string      `json:"status" gorm:"type:varchar(32);default:'unknown'"`
	StatusMsg  string      `json:"statusMsg" gorm:"type:varchar(255);default:''"`
	Version    string      `json:"version" gorm:"type:varchar(64);default:''"`
	LastSeenAt time.Time   `json:"lastSeenAt" gorm:"default:NULL"`
	Summary    NodeSummary `json:"summary" gorm:"serializer:json;type:json"`
}

// NodeSummary 节点摘要快照。
// 只保存最近一次结果（随 Node 行一起存），不做时间序列——历史曲线由各节点自己的 monitor.db 负责，
// 主控只回答“现在哪台有问题”。
type NodeSummary struct {
	Hostname   string  `json:"hostname"`
	OS         string  `json:"os"`
	Uptime     uint64  `json:"uptime"`
	CPUPercent float64 `json:"cpuPercent"`
	CPUTotal   int     `json:"cpuTotal"`
	Load1      float64 `json:"load1"`

	MemPercent float64 `json:"memPercent"`
	MemTotal   uint64  `json:"memTotal"`
	MemUsed    uint64  `json:"memUsed"`

	DiskMaxPercent float64 `json:"diskMaxPercent"`
	DiskMaxPath    string  `json:"diskMaxPath"`

	ContainerRunning  int `json:"containerRunning"`
	ContainerTotal    int `json:"containerTotal"`
	ContainerAbnormal int `json:"containerAbnormal"`

	CertTotal         int `json:"certTotal"`         // 参与统计的证书数量，为 0 时下面两个字段无意义
	CertExpiringCount int `json:"certExpiringCount"` // 30 天内到期（含已过期）的证书数量
	CertMinDays       int `json:"certMinDays"`       // 最近一张证书的剩余天数，负数表示已过期

	Version  string    `json:"version"`
	ShotTime time.Time `json:"shotTime"`
}

File: admin/src/api/modules/node.ts
import http from "@/api"

export interface NodeSummary {
	hostname: string
	os: string
	uptime: number
	cpuPercent: number
	cpuTotal: number
	load1: number
	memPercent: number
	memTotal: number
	memUsed: number
	diskMaxPercent: number
	diskMaxPath: string
	containerRunning: number
	containerTotal: number
	containerAbnormal: number
	certTotal: number
	certExpiringCount: number
	certMinDays: number
	version: string
	shotTime: string
}

export interface NodeWarning {
	type: "offline" | "disk" | "cert" | "container" | "unauthorized"
	level: "warn" | "danger"
	value: number
}

export interface NodeItem {
	id: number
	name: string
	addr: string
	entrance: string
	connectMode: string
	skipVerify: boolean
	isProd: boolean
	sort: number
	status: "online" | "offline" | "unauthorized" | "unknown"
	statusMsg: string
	version: string
	lastSeenAt: string
	summary: NodeSummary
	warnings: NodeWarning[]
	hasToken: boolean
	/** 已保存令牌的明文长度（不含内容）。与 tokenLenExpected 不一致说明存错了 */
	tokenLen: number
	/** 节点签发令牌的标准长度，由后端给出，不写死在前端 */
	tokenLenExpected: number
	/** 是否已配置控制令牌。false 表示该节点只能观测，不能代理操作 */
	hasControlToken: boolean
	controlTokenLen: number
}

export interface NodeSaveParams {
	id?: number
	name: string
	addr: string
	accessToken: string
	/** 控制令牌。留空=不修改（编辑态）；"-" = 清除，即关闭该节点的操作能力 */
	controlToken?: string
	entrance?: string
	skipVerify?: boolean
	isProd?: boolean
	sort?: number
}

export const nodeListAPI = () => {
	return http.get<NodeItem[]>(`/node/list`)
}

export const nodeCreateAPI = (params: NodeSaveParams) => {
	return http.post<any>(`/node/create`, params)
}

export const nodeUpdateAPI = (params: NodeSaveParams) => {
	return http.post<any>(`/node/update`, params)
}

export const nodeDeleteAPI = (params: { id: number }) => {
	return http.post<any>(`/node/del`, params)
}

/** 采集单个节点，返回该节点最新状态 */
export const nodeProbeAPI = (id: number) => {
	return http.post<NodeItem>(`/node/probe/${id}`, {})
}

/** 保存前测试连接，不落库 */
export const nodeProbeDraftAPI = (params: NodeSaveParams) => {
	return http.post<{ addr: string; hostname: string; version: string }>(`/node/probe`, params)
}

/** 手动触发一轮全量采集，返回刷新后的列表 */
export const nodeRefreshAPI = () => {
	return http.post<NodeItem[]>(`/node/refresh`, {})
}

/** 本机是否已开启只读接入 */
export const nodeLocalTokenStatusAPI = () => {
	return http.get<{ enabled: boolean }>(`/node/local/token`)
}

/** 签发本机只读令牌，明文仅本次返回 */
export const nodeLocalTokenIssueAPI = () => {
	return http.post<{ accessToken: string }>(`/node/local/token/issue`, {})
}

export const nodeLocalTokenRevokeAPI = () => {
	return http.post<any>(`/node/local/token/revoke`, {})
}

/** 本机是否已开启控制接入（可被主控代理执行写操作） */
export const nodeLocalControlStatusAPI = () => {
	return http.get<{ enabled: boolean }>(`/node/local/control`)
}

/** 签发本机控制令牌。该令牌等价于本机管理员，明文仅本次返回 */
export const nodeLocalControlIssueAPI = () => {
	return http.post<{ accessToken: string }>(`/node/local/control/issue`, {})
}

export const nodeLocalControlRevokeAPI = () => {
	return http.post<any>(`/node/local/control/revoke`, {})
}
